### ISSUE TITLE
Slim 4 Application throwing a HttpMethodNotAllowed on file upload

### ISSUE THREAD

**Issue by danshatter** [slimphp/Slim#3057]
I created an application using php slim 4. When I submit a form with a file attached to it, It works perfectly on my local machine but when I hosted it on the server, it also works but only on certain conditions.

If the filename has some certain characters, the application throws a 405 HttpMethodNotAllowed exception.

With the same filename, I tested it on previous applications I hosted using php slim 4 as well but they work well on those servers.

I later tried to use slim 4 methodoverridemiddleware by adding an X-Http-Method-Override with a POST value header to force the request method to POST, it doesn't throw the 405 HttpMethodNotAllowed anymore but the request strips off the body of the file.

The name of the file I usually use to test this is 'busola Martins.(3).jpg. When I remove the apostrophe before the b, it works

Please what might be the problem and how may I fix it

**Comment by l0gicgate** [comment]
Would you be able to put a small repository together to reproduce the issue? Without seeing code it's hard to diagnose.

I can tell you that the problem is most likely with middleware ordering. The `MethodOverrideMiddleware` should come first to override the method before routing is performed.

**Comment by danshatter** [comment]
The code actually works. It's just that the server I hosted the application throws the exception when the filename has some special characters like the apostrophe.

It's giving me concern because I can't control what filename users file will have when uploading files

**Comment by JustSteveKing** [comment]
The only reason a Method not allowed exception would be thrown is because you are using the wrong HTTP verb on an endpoint. Make sure that your form is submitting the correct method

**Comment by danshatter** [comment]
Thanks for the reply. I changed the position of the Methodoverridemiddleware but the result was still the same. I was on a deadline so I just had find the list of characters that was causing the errors.

I found out the any file that has apostrophe('), semi-colon(;) and equal sign(=) in its filename throws the exception.

I know it's hard to believe but I've been using slim 4 for a very long time and this is the first time I am experiencing this issue. In my previous projects, it's works great regardless of the filename.

I have come to the conclusion that the problem is from the server I hosted the application, not from slim.

Because I didn't have much time, I just put a write up for users that they should not upload any file that contains those characters above. If they did, I made sure the application would still be give them the chance to upload another picture

I hope I never cross the path of such a server anymore. If you ever experience such, remember it's apostrophe, semi colon and an equal to sign.

Thank you all for the feedback

**Comment by JustSteveKing** [comment]
If it is a shared server then the apache or nginx configuration could be stricter than you might anticipate, meaning an exception is thrown on non utf8 characters - but I have never seen this myself.

If you still struggle, drop a link to a repo or an example code where both middleware, route and controller can be assessed

**Comment by odan** [comment]
It's hard to help without any code snippets. The hoster may block such invalid file names. Have you checked the server error logs to get more details? How do you upload files? Please add more details. Why do you use the MethodOverrideMiddleware for uploads when a simple `POST` method for uploads would be just fine?

**Comment by ddrv** [comment]
@danshatter 

The problem may not be in the Slim, but in the PSR-7/PSR-17 implementation.

**Comment by danshatter** [comment]
Hello everyone. Thank you for all the response. I have found the root of the problem

I created a sub domain on the server and decided that I will use vanilla PHP to test the file upload to see what what response I will get and I got a 403 error Everytime the filename contains semi colon, equal to and quotation marks. Then I decided to changed my search query on Google to see if anyone has ever gotten a 403 on file upload then I found out that it had to do with mod security and disabling it will solve the issue.

I was using namecheap hosting and they don't allow you to change it yourself so I had to contact them to disable the mod security for me. Now the 405 error from slim is not a problem anymore.

I've always know it was from the server. I really grateful for all the support. Thank you all
